drivers/i3c/i3c_common.c

```diff
@@ -402,23 +402,13 @@ static int i3c_bus_setdasa(const struct device *dev,
 			continue;
 		}
 
-		/*
-		 * If there is a desired dynamic address and it is
-		 * not the same as the static address, wait till
-		 * ENTDAA to do address assignment as this is
-		 * no longer SETDASA.
-		 */
-		if ((desc->init_dynamic_addr != 0U) &&
-		    (desc->init_dynamic_addr != desc->static_addr)) {
-			*need_daa = true;
-			continue;
-		}
-
 		LOG_DBG("SETDASA for 0x%x", desc->static_addr);
 
 		ret = i3c_ccc_do_setdasa(desc);
 		if (ret == 0) {
-			desc->dynamic_addr = desc->static_addr;
+			desc->dynamic_addr = (desc->init_dynamic_addr ? desc->init_dynamic_addr
+								      : desc->static_addr);
+			i3c_reattach_i3c_device(dev, desc, desc->static_addr);
```

Comment:
Need to handle when this call returns errors.

Reply:
This was something that I did think about.... but I decided not to do anything if there is an error here as this api is optional and could return ENOSYS..... but there is the issue of if that address is not available.

I do think this reattach api could be made to handle the addr_list before calling in to the driver implemented function.... but that would require some refactoring of how i3c controllers have their config/data commomly defined. I wrote a bit about it in this issue https://github.com/zephyrproject-rtos/zephyr/issues/51828